File: src/main/java/buyhatke/CompoundWord.java
package buyhatke;

import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class CompoundWord {

    List<String> wordDictionary;

    Boolean isCompound(List<String> dictionary, String word) {
        int start = 0;
        int end = word.length();
        boolean available = true;
        int temp = start;
        for (int j = 0; j < dictionary.size(); j++) {
            if (!word.equals(dictionary.get(j))
                    && word.substring(start, end).startsWith(dictionary.get(j))) {
                start = dictionary.get(j).length();
                j = 0;
            }
        }

        return start < end;
    }

    private CompoundWord() {
        wordDictionary = new ArrayList<>();
    }

    public static void main(String args[]) throws Exception {
        Scanner s = new Scanner(System.in);
        CompoundWord problem = new CompoundWord();

        for (int i = 0; i < 20; i++) {
            problem.wordDictionary.add(s.next());
        }

        problem.wordDictionary.sort(String::compareTo);

        for (int i = 0; i < 20; i++) {
            if (problem.isCompound(problem.wordDictionary, problem.wordDictionary.get(i))) {
                System.out.println(problem.wordDictionary.get(i));
            }
        }
    }
}

File: src/main/kotlin/hackerearth/buyhatke/CompoundWord.kt
package hackerearth.buyhatke

import java.util.*
import kotlin.collections.ArrayList

class CompoundWord {

    internal var wordDictionary: MutableList<String> = ArrayList()

    internal fun simplify(compoundWord: String, isRoot: Boolean): List<String>? {
        fun join(first: String, remaining: List<String>): List<String> {
            val result = ArrayList<String>()
            result.add(first)
            remaining.forEach { result.add(it) }
            return result
        }

        if (compoundWord.isEmpty()) {
            return emptyList()
        }

        val options = wordDictionary
                .filter { word -> compoundWord.startsWith(word, ignoreCase = true) }
                .map { word -> simplify(compoundWord.substring(word.length), isRoot = false)?.let { join(word, it) } }
                //.onEach { option -> println(option.toString()) }

        return when {
        // we are inside nested call, return null if no options found
            !isRoot -> options.lastOrNull()

        // only one or no options found
            options.size <= 1 -> listOf(compoundWord)

        // otherwise return second last option, as the last option would be combination of longest words (which is the compound word itself)
            else -> options[options.size - 2] ?: listOf(compoundWord)
        }
    }
}

fun main() {
    val s = Scanner(System.`in`)
    val problem = CompoundWord()

    for (i in 0..19) {
        problem.wordDictionary.add(s.next())
    }

    problem.wordDictionary.sort()

    for (word in problem.wordDictionary) {
        println("$word: ${problem.simplify(word, isRoot = true)}")
    }
}
